docs/link-reference-definition-improvements.md

```diff
@@ -0,0 +1,88 @@
+| **GitHub UI**    | Breaks      | Works          |
+
+So as you can see, we've prioritised GitHub Pages over GitHub Web UI for the time being.
+
+Ideally, we'd like a solution that works with both, but it's not defined yet (see [[link-reference-definitions]] for more details)
+
+#### Workaround
+
+For the time being, you can use relative `[markdown links](markdown-link.md)` syntax.
+
+**Pros:**
+
+- This will work on all platforms.
+
+**Cons:**
+
+- It will break the Markdown Notes [[backlinking]] support
+- Less convenient to write
+
+### Finding certain words clutter the VS Code search results
+
+Since link reference definitions have `[//begin]` and `[//end]` guards with explanatory text that use certain words, these words (like "generate") appear in VS Code search results if you happen to search matching strings from the workspace.
+
+## Improvement Proposal
+
+Problem space in essence:
+
+- During edit-time (when modifying the markdown files in an editor)
+  - link reference definitions are needed if user uses editor extensions that don't understand wikilinks
+  - link reference definitions may be annoying since they 
+    - add content to files that the user hasn't typed in by themselves
+    - get out of date if user uses a tool that doesn't autogenerate them
+    - may clutter the search results
+- During build-time (when converting markdown to html for publishing purposes)
+  - link reference definitions are needed, if the files are published via such tools (or to such platforms) that don't understand wikilinks
+  - link reference definitions might have to be in different formats depending on the publish target (e.g. Github pages vs Github UI)
+
+The potential solution:
+
+- For edit-time
+  - Make edit-time link reference definition generation optional (via user settings and maybe even "off" by default, so that the user wouldn't typically see the auto-generated section in the file at all if it's really not needed)
+  - Make format of the link reference definition configurable (whether to include '.md' or not)
+  - Out of recommended extensions, currently only "markdown links" doesn't support them (?). However even its [code](https://github.com/tchayen/markdown-links/blob/master/src/parsing.ts#L25) seems to include wikilink parser, so it might just be a bug?
+  - If user really want's to use an extension that requires "plain links", they can set the setting "on" and work as in current Foam
+- For build-time
+  - Have publish targets defined in settings, that support both turning the link reference definitions on/off and defining their format (.md or not)
+  - With Foam repo, just use edit-time link reference definitions with '.md' extension - this makes the links work in the Github UI
+  - Have publish target defined for Github pages, that doesn't use '.md' extension, but still has the link reference definitions. Generate the output into gh-pages branch (or separate repo) with automation.
```

Comment:
They should either go to `gh-pages` or `master/docs`, since those are the two supported GH pages default sources.

Reply:
If we are doing a build step anyway, I wonder if we should even use link ref defs, or if we should just nuke 'em, and "transpile" wiki-links to relative markdown links inline? I think the latter, but could have missed a reason why not.

Reply:
edit-time use of extensions not supporting wikilinks would still need the link ref defs, right? Running those against separate build-directory might be something that users wouldn't want, especially if they don't have a need to build (meaning they are not planning to publish their notes). If you mean that we would transpile the "live" md-files from wikilinks to normal md-links, that would IMO result uglier md content and reduce the whole editing experience (so wouldn't do that either).

Reply:
I meant that if we have a structure:

```
/
/file.md
/.build
/.build/file.md
```

Then is there reason to generate link reference definitions into `.build/file.md`, or could we transpile `[[other]]` to `[Other](other.md)` there?

While still generating link reference definitions in `file.md` while editing.

I'm honestly not sure which is better. I feel like transpiling inline would make the output more consumable and portable, since the Markdown spec(s) ecosystem is messy, and not every tool supports link reference definitions but they all support standard links.

Thoughts?

Reply:
If the generated site would contain "see source files" or "download editable copy" (with a requirement of having nice-to-look-at wikilinks in place), having refdefs in files would make publishing this blob easier. Otherwise the publisher would have to make also the original files separately available. Maybe also some metrics could change due to inline transpilation (e.g. amount of characters in an article) which might be confusing when comparing the results between edit-time and after build. How useful / critical these use cases are, I'm not quite sure at this point.

Reply:
took this into consideration in the latest commit